File: backend/mr-db-api/src/main/java/io/featurehub/db/api/PersonApi.java
package io.featurehub.db.api;

import io.featurehub.mr.model.Person;
import io.featurehub.mr.model.PersonType;
import io.featurehub.mr.model.SortOrder;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.List;
import java.util.Set;
import java.util.UUID;

public interface PersonApi {

  @Nullable
  Person update(@NotNull UUID id, @NotNull Person person, @NotNull Opts opts, @NotNull UUID updatedBy) throws OptimisticLockingException;

  // used to determine if the database has no user, which is possible if using external auth
  boolean noUsersExist();

  class PersonPagination {
    public int max;
    public List<Person> people;
    public List<PersonToken> personsWithOutstandingTokens;
    public List<UUID> personIdsWithExpiredTokens;
  }

  class PersonToken {
    public String token;
    public UUID id;

    public PersonToken(String token, UUID id) {
      this.token = token;
      this.id = id;
    }
  }

  class DuplicatePersonException extends Exception {
    public DuplicatePersonException() {
      super();
    }
  }

  @NotNull PersonPagination search(@Nullable String filter, @Nullable SortOrder sortOrder, int offset, int max,
                                   @NotNull Set<PersonType> personTypes,
                                   Opts opts);

  @Nullable Person get(@NotNull UUID id, Opts opts);
  @Nullable Person get(@NotNull String email, Opts opts);

  @Nullable Person getByToken(@NotNull String id, Opts opts);

  @Nullable PersonToken create(@NotNull String email, @Nullable String name, @Nullable UUID createdBy) throws DuplicatePersonException;
  @Nullable CreatedServicePerson createServicePerson(@NotNull String name, @Nullable UUID createdBy);
  @Nullable CreatedServicePerson resetServicePersonToken(@NotNull UUID serviceAccountId);

  boolean delete(@NotNull String email);
}

File: backend/mr-db-api/src/main/kotlin/io/featurehub/db/api/PersonApi.kt
package io.featurehub.db.api

import io.featurehub.mr.model.*
import java.util.*

interface PersonApi {
  @Throws(OptimisticLockingException::class)
  fun update(id: UUID, person: Person, opts: Opts, updatedBy: UUID): Person?

  // used to determine if the database has no user, which is possible if using external auth
  fun noUsersExist(): Boolean
  data class PersonPagination(
    val max: Int,
    val searchPeople: List<SearchPerson>,
    val people: List<Person>,
    val personsWithOutstandingTokens: List<PersonToken>,
    val personIdsWithExpiredTokens: List<UUID>
  )

  data class PersonToken(val token: String, val id: UUID)
  class DuplicatePersonException : Exception()

  fun search(
    filter: String?, sortOrder: SortOrder?, offset: Int, max: Int,
    personTypes: Set<PersonType?>,
    sortBy: SearchPersonSortBy?,
    opts: Opts
  ): PersonPagination

  operator fun get(id: UUID, opts: Opts): Person?
  operator fun get(email: String, opts: Opts): Person?
  fun getByToken(token: String, opts: Opts): Person?

  @Throws(DuplicatePersonException::class)
  fun create(email: String, name: String?, createdBy: UUID?): PersonToken?
  fun createServicePerson(name: String, createdBy: UUID?): CreatedServicePerson?
  fun resetServicePersonToken(serviceAccountId: UUID): CreatedServicePerson?
  fun delete(email: String, deleteGroups: Boolean): Boolean

  fun updateV2(personId: UUID, updatePerson: UpdatePerson, updatedBy: UUID): UUID?
}
